[FLINK-18548][table-planner] support flexible syntax for Temporal table join

leonardBang opened the pull request:
## What is the purpose of the change

* This pull request support flexible syntax for Temporal table join. 
* This PR also can support computed column on temporal table which is a project on TableScan.

## Brief change log

 - Override `convertFrom()` in `SqlToRelConverter` to support flexible `LogicalSnapshot`

## Verifying this change

Add unit tests and ITCase to cover.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

danny0405 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:
```diff
@@ -163,7 +163,35 @@ class FlinkPlannerImpl(
+        // override convertFrom() to support flexible Temporal Table Syntax,
+        // this can be revert once FLINK-16579(Upgrade Calcite version to 1.23) resolved.
+        val relBuilder = config.getRelBuilderFactory.create(cluster, null)
+
```
`val relBuilder: RelBuilder` to eliminate the IDEA warnings.

danny0405 commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/join/LookupJoinITCase.scala:
```diff
@@ -208,15 +247,19 @@ class LookupJoinITCase(legacyTableSource: Boolean, isAsyncMode: Boolean) extends
   def testLeftJoinTemporalTable(): Unit = {
-    val sql = s"SELECT T.id, T.len, D.name, D.age FROM T LEFT JOIN userTable " +
+    if (legacyTableSource) {
+      //Computed column do not support in legacyTableSource.
```
You can use junit `Assume.assumeTrue()` instead.

PS, can we just create another table with computed column there and add test cases for this patch ? It is not necessary to add another test param which makes the test looks verbose and complex.

danny0405 commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/LookupJoinTest.scala:
```diff
@@ -489,9 +493,12 @@ class LookupJoinTest(legacyTableSource: Boolean) extends TableTestBase with Seri
+  @Parameterized.Parameters(name = "LegacyTableSource={0}, useComputedColumn={1}")
   def parameters(): JCollection[Array[Object]] = {
-    Seq[Array[AnyRef]](Array(JBoolean.TRUE), Array(JBoolean.FALSE))
+    Seq[Array[AnyRef]](
```
No need to add another param `useComputedColumn` i think. Just add new tables and new cases you want to test.

leonardBang commented:
@danny0405 I have addressed your comments, Could you have a more look?

leonardBang commented:
@danny0405 could you take a look? thanks..

danny0405 reviewed (approved):
+1, looks good from my side.

leonardBang commented:
cc: @wuchong Could you take a look and help merge? Thanks

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:
```diff
@@ -163,7 +163,35 @@ class FlinkPlannerImpl(
+        // this can be revert once FLINK-16579(Upgrade Calcite version to 1.23) resolved.
+        val relBuilder = config.getRelBuilderFactory.create(cluster, null)
+
+        override def convertFrom(bb: SqlToRelConverter#Blackboard, from: SqlNode): Unit = {
```
Create a JIRA issue to remove this overriding once we bump up Calcite version. And add comment above this method with the JIRA id.

wuchong reviewed (approved):
LGTM.

wuchong commented:
The compile is failed, please fix it first. 

```
FlinkPlannerImpl.scala message=File line length exceeds 100 characters line=168
```

leonardBang commented:
> The compile is failed, please fix it first.
> 
> ```
> FlinkPlannerImpl.scala message=File line length exceeds 100 characters line=168
> ```

I've fixed and rebased

wuchong commented:
Thanks for the updating @leonardBang . 

Will merge this.
